samcli/lib/list/testable_resources/testable_resources_producer.py

```diff
@@ -0,0 +1,367 @@
+            The identifier or physical ID
+
+        Returns
+        -------
+        furl: str
+            The function url in the form of a string
+        """
+        try:
+            response = self.cloudcontrol_client.get_resource(TypeName="AWS::Lambda::Url", Identifier=identifier)
+            if not response.get("ResourceDescription", {}).get("Properties", {}):
+                return "-"
+            response_dict = json.loads(response.get("ResourceDescription", {}).get("Properties", {}))
+            furl = response_dict.get("FunctionUrl", "-")
+            return furl
+        except ClientError as e:
+            if get_client_error_code(e) == "ResourceNotFoundException":
+                return "-"
+            LOG.error("ClientError Exception : %s", str(e))
+            raise SamListUnknownClientError(msg=str(e)) from e
+
+    def get_stage_list(self, api_id: str, api_type: APIGatewayEnum) -> List[Any]:
+        """
+        Gets a list of stages for a given api of type AWS::ApiGateway::RestApi or AWS::ApiGatewayV2::Api
+
+        Parameters
+        ----------
+        api_id: str
+            The api id or rest api id of the api
+        api_type: APIGatewayEnum
+            The type of api, AWS::ApiGateway::RestApi or AWS::ApiGatewayV2::Api
+
+        Returns
+        -------
+        response_list: List[Any]
+            A list of stages for the api
+        """
+        response_list: List[Any]
+        try:
+            response_list = []
+            response: dict
+            search_key: str
+            stage_name_key: str
+            if api_type == APIGatewayEnum.API_GATEWAY:
+                response = self.apigateway_client.get_stages(restApiId=api_id)
+                search_key = "item"
+                stage_name_key = "stageName"
+            else:
+                response = self.apigatewayv2_client.get_stages(ApiId=api_id)
+                search_key = "Items"
+                stage_name_key = "StageName"
+            if not response.get(search_key, []):
+                return response_list
+            for item in response.get(search_key, []):
+                if item.get(stage_name_key, None):
+                    response_list.append(item.get(stage_name_key, ""))
+            return response_list
+        except ClientError as e:
+            if get_client_error_code(e) == "NotFoundException":
+                return []
+            LOG.error("ClientError Exception : %s", str(e))
+            raise SamListUnknownClientError(msg=str(e)) from e
+        except BotoCoreError as e:
+            LOG.error("Botocore Exception : %s", str(e))
+            raise SamListUnknownBotoCoreError(msg=str(e)) from e
+
+    def build_api_gw_endpoints(self, physical_id: str, stages: list) -> list:
+        api_list = []
+        for stage in stages:
+
+            api_list.append(f"https://{physical_id}.execute-api.{self.region}.amazonaws.com/{stage}")
+        return api_list
+
+    def produce(self):
```

Comment:
This function is quite large and does multiple things. Is there any way we can break it up a bit? Similar to how you moved out the `get_local_testable_resources()` function.

Reply:
Moved out the cloud testable resources code